homeassistant/components/ipp/sensor.py

```diff
@@ -0,0 +1,178 @@
+            name=name,
+            icon=icon,
+            enabled_default=enabled_default,
+        )
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return f"{self.coordinator.data.info.uuid}_{self._key}"
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit this state is expressed in."""
+        return self._unit_of_measurement
+
+
+class IPPMarkerSensor(IPPSensor):
+    """Defines an IPP marker sensor."""
+
+    def __init__(
+        self, entry_id: str, coordinator: IPPDataUpdateCoordinator, marker_index: int
+    ) -> None:
+        """Initialize IPP marker sensor."""
+        self.marker_index = marker_index
+
+        super().__init__(
+            coordinator=coordinator,
+            entry_id=entry_id,
+            icon="mdi:water",
+            key=f"marker_{marker_index}",
+            name=f"{coordinator.data.info.name} {coordinator.data.markers[marker_index].name}",
+            unit_of_measurement=UNIT_PERCENTAGE,
+        )
+
+    @property
+    def device_state_attributes(self) -> Optional[Dict[str, Any]]:
+        """Return the state attributes of the entity."""
+        return {
+            ATTR_MARKER_HIGH_LEVEL: self.coordinator.data.markers[
+                self.marker_index
+            ].high_level,
+            ATTR_MARKER_LOW_LEVEL: self.coordinator.data.markers[
+                self.marker_index
+            ].low_level,
+            ATTR_MARKER_TYPE: self.coordinator.data.markers[
+                self.marker_index
+            ].marker_type,
+        }
+
+    @property
+    def state(self) -> Union[None, str, int, float]:
+        """Return the state of the sensor."""
+        return self.coordinator.data.markers[self.marker_index].level
+
+
+class IPPPrinterSensor(IPPSensor):
+    """Defines an IPP printer sensor."""
+
+    def __init__(self, entry_id: str, coordinator: IPPDataUpdateCoordinator) -> None:
+        """Initialize IPP printer sensor."""
+        super().__init__(
+            coordinator=coordinator,
+            entry_id=entry_id,
+            icon="mdi:printer",
+            key="printer",
+            name=coordinator.data.info.name,
+            unit_of_measurement=None,
+        )
+
+    @property
+    def device_state_attributes(self) -> Optional[Dict[str, Any]]:
+        """Return the state attributes of the entity."""
+        return {
+            ATTR_PRINTER_INFO: self.coordinator.data.info.printer_info,
+            ATTR_PRINTER_SERIAL: self.coordinator.data.info.serial,
+            ATTR_PRINTER_LOCATION: self.coordinator.data.info.location,
+            ATTR_PRINTER_STATE_MESSAGE: self.coordinator.data.state.message,
+            ATTR_PRINTER_STATE_REASON: self.coordinator.data.state.reasons,
+            ATTR_PRINTER_COMMAND_SET: self.coordinator.data.info.command_set,
+            ATTR_PRINTER_URI_SUPPORTED: self.coordinator.data.info.printer_uri_supported,
```

Comment:
Serial is part of the device info, we don't need to include that in the attributes. What is in info? And what is uri supported ?

Reply:
The serial is not always available as it's part of a device id string that has to be parsed and we already have a unique uuid that is always set.

is it possible to define multiple identifiers?

Reply:
printer-uri-supported (1setOf uri)

   This REQUIRED Printer attribute contains at least one URI for the
   Printer object.  It OPTIONALLY contains more than one URI for the
   Printer object.    An administrator determines a Printer object's
   URI(s) and configures this attribute to contain those URIs by some
   means outside the scope of this IPP/1.1 document.  The precise format
   of this URI is implementation dependent and depends on the protocol.

Reply:
printer-info (text(127))

   This Printer attribute identifies the descriptive information about
   this Printer object.  This could include things like: "This printer
   can be used for printing color transparencies for HR presentations",
   or "Out of courtesy for others, please print only small (1-5 page)
   jobs at this printer", or even "This printer is going away on July 1,
   1997, please find a new printer".